system/lib/pthread/emscripten_futex_wait.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+#include <assert.h>
+#include <errno.h>
+#include <emscripten/threading.h>
+
+int _emscripten_futex_wait_non_blocking(volatile void *addr, uint32_t val, double max_wait_ms);
+
+int emscripten_futex_wait(volatile void *addr, uint32_t val, double max_wait_ms) {
+  if ((((intptr_t)addr)&3) != 0) {
+    return -EINVAL;
+  }
+
+  int ret;
+  emscripten_conditional_set_current_thread_status(EM_THREAD_STATUS_RUNNING, EM_THREAD_STATUS_WAITFUTEX);
+
+  if (!emscripten_thread_can_block()) {
+    ret = _emscripten_futex_wait_non_blocking(addr, val, max_wait_ms);
+    emscripten_conditional_set_current_thread_status(EM_THREAD_STATUS_WAITFUTEX, EM_THREAD_STATUS_RUNNING);
+    return ret;
+  }
+
+  int64_t max_wait_ns = (int64_t)(max_wait_ms*1000*1000);
+  ret = __builtin_wasm_memory_atomic_wait32((int*)addr, val, max_wait_ns);
```

Comment:
Should we remap `max_wait_ns` to `-1` when a infinite timeout (`INFINITY`) is specified?

Reply:
Ah, yes that is one of the differences between the native and JS versions.  Good catch.